pkg/instancetype/controller/vm/controller_test.go

```diff
@@ -0,0 +1,1052 @@
+
+			_, err := instancetypeController.Sync(vm, vmi)
+			Expect(err).To(MatchError(ContainSubstring("not found")))
+			testutils.ExpectEvents(recorder, common.FailedCreateVirtualMachineReason)
+		})
+
+		It("should fail to sync if a VirtualMachineClusterInstancetype cannot be found", func() {
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name: "foobar",
+				Kind: instancetypeapi.ClusterSingularResourceName,
+			}
+
+			_, err := instancetypeController.Sync(vm, vmi)
+			Expect(err).To(MatchError(ContainSubstring("not found")))
+			testutils.ExpectEvents(recorder, common.FailedCreateVirtualMachineReason)
+		})
+
+		It("should fail to sync if the VirtualMachineInstancetype conflicts with the VirtualMachineInstance", func() {
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name: instancetypeObj.Name,
+				Kind: instancetypeapi.SingularResourceName,
+			}
+
+			vm.Spec.Template.Spec.Domain.CPU = &virtv1.CPU{
+				Sockets: uint32(1),
+				Cores:   uint32(4),
+				Threads: uint32(1),
+			}
+
+			_, err := instancetypeController.Sync(vm, vmi)
+			Expect(err).To(MatchError(ContainSubstring("conflicts with selected instance type")))
+			testutils.ExpectEvents(recorder, common.FailedCreateVirtualMachineReason)
+		})
+
+		It("should fail to sync if an existing ControllerRevision is found with unexpected VirtualMachineInstancetypeSpec data", func() {
+			unexpectedInstancetype := instancetypeObj.DeepCopy()
+			unexpectedInstancetype.Spec.CPU.Guest = 15
+
+			instancetypeRevision, err := instancetype.CreateControllerRevision(vm, unexpectedInstancetype)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+				context.Background(), instancetypeRevision, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name: instancetypeObj.Name,
+				Kind: instancetypeapi.SingularResourceName,
+			}
+
+			_, err = instancetypeController.Sync(vm, vmi)
+			Expect(err).To(MatchError(ContainSubstring("found existing ControllerRevision with unexpected data")))
+			testutils.ExpectEvents(recorder, common.FailedCreateVirtualMachineReason)
+		})
+	})
+
+	Context("preference", func() {
+		var clusterPreference *v1beta1.VirtualMachineClusterPreference
+
+		BeforeEach(func() {
+			clusterPreference = &v1beta1.VirtualMachineClusterPreference{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "clusterPreference",
+					UID:        resourceUID,
+					Generation: resourceGeneration,
+				},
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: v1beta1.SchemeGroupVersion.String(),
+					Kind:       "VirtualMachineClusterPreference",
+				},
+				Spec: v1beta1.VirtualMachinePreferenceSpec{
+					Firmware: &v1beta1.FirmwarePreferences{
+						DeprecatedPreferredUseEfi: pointer.P(true),
+					},
+					Devices: &v1beta1.DevicePreferences{
+						PreferredDiskBus:        virtv1.DiskBusVirtio,
+						PreferredInterfaceModel: "virtio",
```

Comment:
Is there a const for it?

Reply:
Not yet, easy to add however.

https://github.com/kubevirt/kubevirt/blob/48a413d72b018a7f9503ed11c7dbb76cf8624e27/staging/src/kubevirt.io/api/core/v1/schema.go#L1246-L1250